common/city.cpp

```diff
@@ -2876,6 +2879,9 @@ void set_city_production(struct city *pcity,
    * on, so shield waste will include shield bonuses. */
   output_type_iterate(o)
   {
+    // If there is a cache it must not be empty.
+    fc_assert(!pcwaste || !pcwaste->empty());
```

Comment:
This should probably go before the loop to avoid getting 6 messages every time. I think returning if the assert fails is fine, so use `fc_assert_ret(condition)`.

Reply:
Same here.